src/ir/possible-contents.h

```diff
@@ -0,0 +1,577 @@
+//                     TODO: Add ConeType, which would include subtypes.
+//                     TODO: Add ExactTypePlusContents or such, which would be
+//                           used on e.g. a struct.new with an immutable field
+//                           to which we assign a constant: not only do we know
+//                           the exact type, but also certain field's values.
+//
+//  * Many:            Anything else. Many things are possible here, and we do
+//                     not track what they might be, so we must assume the worst
+//                     in the calling code.
+//
+class PossibleContents {
+  struct None : public std::monostate {};
+
+  struct GlobalInfo {
+    Name name;
+    // The type of the global in the module. We stash this here so that we do
+    // not need to pass around a module all the time.
+    // TODO: could we save size in this variant if we did pass around the
+    //       module?
+    Type type;
+    bool operator==(const GlobalInfo& other) const {
+      return name == other.name && type == other.type;
+    }
+  };
+
+  using ExactType = Type;
+
+  struct Many : public std::monostate {};
+
+  // TODO: This is similar to the variant in PossibleConstantValues, and perhaps
+  //       we could share code, but extending a variant using template magic may
+  //       not be worthwhile. Another option might be to make PCV inherit from
+  //       this and disallow ExactType etc., but PCV might get slower.
+  using Variant = std::variant<None, Literal, GlobalInfo, ExactType, Many>;
+  Variant value;
+
+public:
+  PossibleContents() : value(None()) {}
+  PossibleContents(const PossibleContents& other) : value(other.value) {}
+
+  // Most users will use one of the following static functions to construct a
+  // new instance:
+
+  static PossibleContents none() {
+    PossibleContents ret;
+    ret.value = None();
+    return ret;
+  }
+  static PossibleContents literal(Literal c) {
+    PossibleContents ret;
+    ret.value = c;
+    return ret;
+  }
+  static PossibleContents global(Name name, Type type) {
+    PossibleContents ret;
+    ret.value = GlobalInfo{name, type};
+    return ret;
+  }
+  static PossibleContents exactType(Type type) {
+    PossibleContents ret;
+    ret.value = ExactType(type);
+    return ret;
+  }
+  static PossibleContents many() {
+    PossibleContents ret;
+    ret.value = Many();
+    return ret;
+  }
+
+  PossibleContents& operator=(const PossibleContents& other) {
+    value = other.value;
+    return *this;
+  }
+
+  bool operator==(const PossibleContents& other) const {
+    return value == other.value;
+  }
+
+  bool operator!=(const PossibleContents& other) const {
+    return !(*this == other);
+  }
+
+  // Combine the information in a given PossibleContents to this one. The
+  // contents here will then include whatever content was possible in |other|.
+  void combine(const PossibleContents& other) {
```

Comment:
I think it would make sense to move this out of the header and into the .cpp file.

Reply:
This is by far the most performance-sensitive function in this class. It is hammered on constantly by the flow analysis. So I think it is important to keep it inlineable?

Reply:
(though if we start to use PossibleContents in more places, that might change... but we can refactor later if so)

Reply:
Oh right, I missed that. Thanks for investigating!